compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp

```diff
@@ -438,14 +516,234 @@ void distributeTransferWrites(vector::TransferWriteOp writeOp,
+}
+
+static Value createReductionOp(vector::CombiningKind combiningKind, Value lhs,
+                               Value rhs, Location loc, OpBuilder &rewriter) {
+  switch (combiningKind) {
+    case vector::CombiningKind::ADD:
+      return rewriter.create<arith::AddFOp>(loc, lhs, rhs);
+    case vector::CombiningKind::MAXF:
+      return rewriter.create<arith::MaxFOp>(loc, lhs, rhs);
+    default:
+      assert("Unsupported vector combining kind!");
+      break;
+  }
+  return Value{};
+}
+
+using Triplet = std::tuple<vector::MultiDimReductionOp, vector::BroadcastOp,
+                           vector::TransposeOp>;
+using bodyType = std::function<void(std::array<int, DimType::NumDims> &)>;
+
+template <std::size_t T>
+void iterate(int dimType, std::array<int, T> &order,
+             std::array<int, DimType::NumDims> &state, const Layout &layout,
+             bodyType body) {
+  if (dimType == DimType::NumDims) {
+    body(state);
+    return;
+  }
+  if ((std::find(order.begin(), order.end(), dimType) != order.end()) &&
+      (!isLaneId(dimType))) {
+    for (int i = 0; i < layout.shape[dimType]; i++) {
+      state[dimType] = i;
+      iterate(dimType + 1, order, state, layout, body);
+    }
+  } else {
+    iterate(dimType + 1, order, state, layout, body);
+  }
+}
+
+void distributeReductionBroadcastTranspose(
+    vector::MultiDimReductionOp reductionOp, DenseMap<Value, Layout> &layoutMap,
+    DenseMap<Value, Value> &simdToSimtMap, OpBuilder &rewriter,
+    SmallVectorImpl<Triplet> &triplets) {
+  vector::BroadcastOp broadcastOp{nullptr};
+  vector::TransposeOp transposeOp{nullptr};
+  for (Operation *user : reductionOp.getResult().getUsers()) {
+    if (auto broadcast = dyn_cast<vector::BroadcastOp>(user)) {
+      for (Operation *bUser : broadcast.getResult().getUsers()) {
+        if (auto transpose = dyn_cast<vector::TransposeOp>(bUser)) {
+          transposeOp = transpose;
+          break;
+        }
+      }
+      broadcastOp = broadcast;
+      break;
+    }
+  }
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(reductionOp);
+  Value source = reductionOp.getSource();
+  Type elementType = source.getType().cast<ShapedType>().getElementType();
+  if (!layoutMap.count(source)) return;
+  if (!simdToSimtMap.count(source)) return;
+  if (!broadcastOp) return;
+  Location loc = reductionOp.getLoc();
+  Layout layout = layoutMap.at(source);
+  auto reductionDims = llvm::to_vector<4>(
+      reductionOp.getReductionDims().getAsRange<IntegerAttr>());
+  vector::CombiningKind combiningKind = reductionOp.getKind();
+  // Only support reduction on one dimension
+  if (reductionDims.size() > 1) return;
+  int reductionDim = reductionDims[0].getInt();
+  std::array<int, 4> reductionOrder = layout.order[reductionDim];
+  std::array<int, 4> parallelOrder = layout.order[!reductionDim];
+  Value acc = reductionOp.getAcc();
+  APFloat floatValue(0.0);
+  if (!matchPattern(acc, m_ConstantFloat(&floatValue))) return;
```

Comment:
Maybe add a TODO here as we should be able to handle any accumulator without too much problems.

Reply:
Sure will add that